### PR TITLE
changes based on code reviews

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
- moved kvstore stype checks to backend and added test for invalid stype pulls
- moved scipy import inside test util function
- moved benchmark files in /sparse folder

@piiswrong @reminisce

### CODE REVIEW

File: src/kvstore/kvstore_local.h [hunk 1]
```diff
@@ -174,13 +175,75 @@ class KVStoreLocal : public KVStore {
 
  protected:
   /**
-   * \brief group values on keys
+   * \brief group values on keys for push
    */
-  template <typename V>
+  void GroupKVPairsPush(const std::vector<int>& keys,
```
Comment (reminisce) [inline src/kvstore/kvstore_local.h]: A second thought is defining validator lambda function in the caller functions or a common place accessible to various callers and pass it to `GroupKVPairs`. In this way, there is no need to define extra interfaces of GroupKVPairsXXXX where XXXX stands for Push and Pull, respectively. Is this feasible?

Comment (eric-haibin-lin) [inline src/kvstore/kvstore_local.h]: I also thought about that. What file do you think is the best place to put the free function? I started with `GroupKVPairsForPushRsp` but it is too long (>100 char) causing lint to fail..

File: src/kvstore/kvstore_local.h [hunk 2]
```diff
@@ -174,13 +175,75 @@ class KVStoreLocal : public KVStore {
...
+  }
+  /**
+   * \brief group values on keys for pull
+   */
+  void GroupKVPairsPull(const std::vector<int>& keys,
+                        const std::vector<NDArray*>& values,
+                        std::vector<int> *uniq_keys,
+                        std::vector<std::vector<NDArray*>> *grouped_vals) {
+    // check if the storage type of a value is valid
+    auto validator = [this](const int key, NDArray* nd) -> bool {
```
Comment (reminisce) [inline src/kvstore/kvstore_local.h]: `const NDArray*`

File: src/kvstore/kvstore_local.h [hunk 3]
```diff
@@ -174,13 +175,75 @@ class KVStoreLocal : public KVStore {
...
+  }
+  /**
+   * \brief group values on keys for row_sparse_pull
+   */
+  void GroupKVPairsPullRsp(const std::vector<int>& keys,
+                           const std::vector<std::pair<NDArray*, NDArray>>& values,
+                           std::vector<int> *uniq_keys,
+                           std::vector<std::vector<std::pair<NDArray*, NDArray>>> *grouped_vals) {
+    // check if the storage type of a value is valid
+    auto validator = [this](const int key, std::pair<NDArray*, NDArray> val_rowid) -> bool {
```
Comment (reminisce) [inline src/kvstore/kvstore_local.h]: `const std::pair<NDArray*, NDArray>& val_rowid`

**Comment by eric-haibin-lin** [comment]
Added const to lamdas

**Pull request merged**